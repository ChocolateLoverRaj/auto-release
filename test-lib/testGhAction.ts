import { spawn } from 'child_process'
import streamToString from 'stream-to-string'
import { once } from 'events'
import { EOL } from 'os'
import { join } from 'path'
import { writeFile } from 'jsonfile'
import defaults from 'object.defaults'
import getPort from 'get-port'
import express from 'express'
import { createServer } from 'http'

interface Result {
  outputs: Record<string, string>
  stdout: string
}

interface Commit {
  message: string
}

interface PullRequest {
  commits: Commit[]
}

interface Repo {
  pullRequests: Record<number, PullRequest>
  token?: string
}

interface PrEvent {
  number: number
}

interface Options {
  inputs: Record<string, string>
  event: PrEvent
  repo: Repo
  env: Record<string, string>
}

const testGhAction = async (file: string, partialOptions: Partial<Options> = {}): Promise<Result> => {
  const options: Options = defaults(partialOptions, { inputs: [], event: { number: 1 }, repo: { pullRequests: {} } })
  const { event, inputs, repo } = options
  const eventFile = join(__dirname, '../test-tmp/event.json')
  const port = await getPort()

  const server = express()
  const fakeGhServer = createServer(server).listen(port)

  server.get('/repos/ChocolateLoverRaj/test-semver/pulls/:number/commits', (req, res) => {
    const n = parseInt(req.params.number)
    const pr = repo.pullRequests[n]
    if (pr === undefined) {
      res.sendStatus(404)
      return
    }
    res.json(pr.commits.map(commit => ({ commit })))
  })

  await Promise.all([
    writeFile(eventFile, {
      pull_request: {
        commits: repo.pullRequests[event.number].commits.length,
        number: event.number
      },
      repo: {
        name: 'test-semver',
        owner: {
          login: 'ChocolateLoverRaj'
        }
      }
    }, { spaces: 2 }),
    once(fakeGhServer, 'listening')
  ])

  const c = spawn('node', [file], {
    env: {
      ...process.env,
      GITHUB_EVENT_PATH: eventFile,
      GITHUB_API_URL: `http://localhost:${port}`,
      ...Object.fromEntries(Object.entries(inputs).map(([k, v]) => [`INPUT_${k.toUpperCase()}`, v])),
      ...options.env
    }
  })
  const stdoutStrPromise = streamToString(c.stdout)
  const stderrStr = streamToString(c.stderr)
  const [code] = await once(c, 'exit')
  fakeGhServer.close()
  if (code !== 0) {
    throw new Error(
      `Process exited with code ${code as number}\n` +
      'Stderr is below:\n' +
      await stderrStr
    )
  }
  const setOutputStr = '::set-output name='
  const stdoutStr = (await stdoutStrPromise)
  return {
    outputs: Object.fromEntries(stdoutStr
      .split(EOL)
      .map(str => {
        if (str.startsWith(setOutputStr)) {
          const keyValue = str.slice(setOutputStr.length).split('::')
          if (keyValue.length !== 2) throw new Error('Bad set output syntax')
          return keyValue
        }
        return undefined
      })
      .filter(v => v !== undefined) as any
    ),
    stdout: stdoutStr
  }
}

export default testGhAction
