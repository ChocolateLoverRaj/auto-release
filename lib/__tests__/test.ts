import testGhAction from '../../test-lib/testGhAction'
import { join } from 'path'
import PullRequest from '../../test-lib/PullRequest'
import assignSame from '../../test-lib/assignSame'

const mainFilePath = join(__dirname, '../../dist/index.js')

test('pull request', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: assignSame(new PullRequest(), {
          commits: [{
            message: 'Feature: add triangles'
          }, {
            message: 'Breaking: replaced triangle with polygon'
          }, {
            message: 'Fix: polygons have minimum of 3 sides'
          }, {
            message: 'Chore: updated GitHub action script'
          }]
        })
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('major')
})

test('push', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      commits: [{
        message: 'Feat: add images'
      }, {
        message: 'Fix: fix bug about images'
      }]
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('minor')
})

test('PR merge', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      commits: [{
        message: 'Merge pull request #24 from ChocolateLoverRaj/test-semver'
      }, {
        message: 'Feat: allow case insensitive words'
      }]
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('minor')
})

test('branch merge', async () => {
  const { stdout, outputs } = await testGhAction(mainFilePath, {
    event: {
      commits: [{
        message: "Merge branch 'main' into eslint-tsconfig-no-tests"
      }, {
        message: 'Chore: CI stuff'
      }]
    }
  })
  expect(stdout).toMatchSnapshot()
  expect(outputs.increment).toBe('none')
})

test('invalid commit message', async () => {
  await expect(testGhAction(join(__dirname, '../../dist/index.js'), {
    event: {
      number: 1
    },
    repo: {
      pullRequests: {
        1: assignSame(new PullRequest(), {
          commits: [{ message: 'made it better' }]
        })
      },
      token: 'token'
    },
    env: {
      GITHUB_TOKEN: 'token'
    }
  })).rejects.toMatchSnapshot()
})

test('scoped commits', async () => {
  const { outputs, stdout } = await testGhAction(mainFilePath, {
    event: {
      commits: [{
        message: 'breaking(react-json-input, antd): drop node 10 support'
      }]
    }
  })
  expect(outputs).toEqual({
    increment: 'major',
    increments_by_scope: JSON.stringify({ 'react-json-input': 'major', antd: 'major' })
  })
  expect(stdout).toMatchSnapshot()
})
