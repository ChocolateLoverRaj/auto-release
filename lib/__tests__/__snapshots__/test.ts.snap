// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PR merge 1`] = `
"Parsing commit messages
Message: Merge pull request #24 from ChocolateLoverRaj/test-semver. Increment: none (pull request merge commit). Scopes: *none*.
Message: Feat: allow case insensitive words. Increment: minor. Scopes: *none*.

::set-output name=increment::minor

::set-output name=increments_by_scope::{}
Largest increment: minor
"
`;

exports[`branch merge 1`] = `
"Parsing commit messages
Message: Merge branch 'main' into eslint-tsconfig-no-tests. Increment: none (merge commit). Scopes: *none*.
Message: Chore: CI stuff. Increment: none. Scopes: *none*.

::set-output name=increment::none

::set-output name=increments_by_scope::{}
Largest increment: none
"
`;

exports[`invalid commit message 1`] = `
[Error: Process exited with code 1
Stderr is below:
Invalid Commit Message: made it better
Messages should follow: <type>[optional scope]: <description>
]
`;

exports[`pull request 1`] = `
"Fetching 4 commits.
Parsing commit messages
Message: Feature: add triangles. Increment: minor. Scopes: *none*.
Message: Breaking: replaced triangle with polygon. Increment: major. Scopes: *none*.
Message: Fix: polygons have minimum of 3 sides. Increment: patch. Scopes: *none*.
Message: Chore: updated GitHub action script. Increment: none. Scopes: *none*.

::set-output name=increment::major

::set-output name=increments_by_scope::{}
Largest increment: major
"
`;

exports[`push 1`] = `
"Parsing commit messages
Message: Feat: add images. Increment: minor. Scopes: *none*.
Message: Fix: fix bug about images. Increment: patch. Scopes: *none*.

::set-output name=increment::minor

::set-output name=increments_by_scope::{}
Largest increment: minor
"
`;

exports[`scoped commits 1`] = `
"Parsing commit messages
Message: breaking(react-json-input, antd): drop node 10 support. Increment: major. Scopes: react-json-input, antd.

::set-output name=increment::major

::set-output name=increments_by_scope::{\\"react-json-input\\":\\"major\\",\\"antd\\":\\"major\\"}
Largest increment for scope 'react-json-input': major
Largest increment for scope 'antd': major
"
`;
